csrc/scheduler/utils.cpp

```diff
@@ -170,21 +170,36 @@ std::optional<size_t> mergeDims(
   if (to_merge.size() == 1) {
     return to_merge[0];
   }
-  std::sort(to_merge.begin(), to_merge.end());
-  size_t left = to_merge[0];
-  int64_t offset = 0;
-  for (auto right = to_merge.begin() + 1; right != to_merge.end(); right++) {
-    auto actual_right = offset-- + *right;
-    tv->merge((int)left, (int)actual_right);
-    for (auto& i : to_update) {
-      if (i == actual_right) {
-        i = left;
-      } else if (i > actual_right) {
-        i--;
+  auto inner = to_merge[0];
+
+  // NOTE: we need to merge dimensions in the order specified by `to_merge`,
+  // otherwise this could results in misaligned memory access due to indexing.
+  // This is because we compute vectorization width before applying scheduling
+  // transformations.
+  for (size_t i = 1; i < to_merge.size(); i++) {
+    auto outer = to_merge[i];
+    if (outer > inner) {
+      tv->reorder({{inner, outer}, {outer, inner}});
+      std::swap(inner, outer);
+    }
```

Comment:
If a `to_merge` has `{1, 0}`, shouldn't we just do `merge(0, 1)`? Otherwise, it seems to contradict with the comment above?

Reply:
So dimensions in `to_merge` is ordered from inner to outer. When we have the next element in `to_merge` that is larger then `inner`, i.e.
```
tv[i0, i1]
outer = 1
inner = 0
```
We would want to have the merge result be tv[i1 * i0].

If we just do `tv->merge(0, 1)`, that will give us `tv[i0 * i1]`. Hence the reordering.

Reply:
Let me put the comment in the code as well

Reply:
Maybe my confusion is, when you say:

> So dimensions in to_merge is ordered from inner to outer

Is the ordering of the inner and outer referring to the ordering in the domain? Or does it refer to the ordering in merge?

Reply:
This is indeed very confusing :confused: 

When I talk about `inner to outer`, I was referring to the order in the `merged iter domain`.
so we want the end result to be something like
`tv->axis(to_merge[i-1]) * tv->axis(to_merge[i-2]) * ... * tv->axis(to_merge[0])`

^^^ this part is what inner vs outer.

In the mean time, when we merge `to_merge[k]` with `to_merge[k-1]`, there's no guarantee that the two item would be in any order. i.e. if to_merge[k] is larger than `to_merge[k-1]`, we need to reorder it, so we'll have the merged result be in `to_merge[k] * to_merge[k-1]` instead of the other way around.

Reply:
actually, I think the last merged dimension doesn't seem to be at the right position... hmmm... interesting...

Reply:
so this is still functional, but it's unsatisfying that the merged dimension would just be the outer most in the original leaf domain. `*std::min_element(to_merge.begin(), to_merge.end())` instead of the logical order.

Maybe I missed out how to do that properly in merge/reorder.

Reply:
I tried to put more comment, feel free to yell at me if it's not helping :stuck_out_tongue: